pkg/iam/mapping.go

```diff
@@ -1,17 +1,117 @@
 package iam
 
-import "errors"
+import (
+	"errors"
+	"fmt"
+)
 
-// Identity represents an IAM identity.
-type Identity struct {
-	Username string   `json:"username"`
-	Groups   []string `json:"groups"`
+var (
+	// ErrNeitherUserNorRole is the error returned when an identity is missing both UserARN
+	// and RoleARN.
+	ErrNeitherUserNorRole = errors.New("arn is neither user nor role")
+
+	// ErrNoKubernetesIdentity is the error returned when an identity has neither a Kubernetes
+	// username nor a list of groups.
+	ErrNoKubernetesIdentity = errors.New("neither username nor groups is set for iam identity")
+)
+
+// Identity represents an IAM identity and its corresponding Kubernetes identity
+type Identity interface {
+	GetARN() string
```

Comment:
Not a fan of the `Get` prefixes here but `Username` and `Groups` will conflict. I'm OK with leaving the `GetARN()` for consistency (at least within the struct) but also don't oppose removing it.

An alternative would be to unexport fields and implement `MarshalJSON()` where one uses a copy of the struct with exported fields but that's also quite verbose.

I'm fine as is.

Reply:
Yeah, I don't like it either. What if I make the structs have different field names:

```
type RoleIdentity struct {
  RoleARN,
  {
    KubernetesUsername string `json:"username", omitempty`,
    KubernetesGroups []string `json:"groups", omitempty`,
  }
}

type Identity interface {
  ARN()
  Username()
  Groups()
}
```

Reply:
i like it!

Reply:
I refactored it (and rebased)